01. Importing libraries

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

02. Importing data

In [45]:
# Import data
path = r'C:\Users\polly\Documents\Instacart Basket Analysis'
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = 0)

03. Creating a test dataframe

In [4]:
# Create a dataframe
df_test = pd.DataFrame()

In [5]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [6]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [7]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [8]:
# Convert data type of column to string
df_test['mix'] = df_test['mix'].astype('str')

4. Finding missing values

In [9]:
# Find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
# Create new dataframe with missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [11]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [12]:
df_prods.shape

(49693, 5)

In [13]:
# Create a new dataframe
df_prods_clean = pd.DataFrame()

In [14]:
# Map new dataframe to non-missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
df_prods_clean.shape

(49677, 5)

5. Finding duplicates

In [16]:
# create new dataframe with duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [18]:
df_prods_clean.shape

(49677, 5)

In [19]:
# Create new dataframe with unique rows
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [20]:
df_prods_clean_no_dups.shape

(49672, 5)

In [21]:
# Export df_prods_clean_no_dups to 'Prepared Data' folder
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

6. Task 4.5

In [22]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [23]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

Question 2 Answer: The max for the 'prices' column looks off and should be investigated further

In [24]:
# Question 3: check for mixed data types in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

Question 3 Answer: There are no mixed-type data in df_ords dataframe, as nothing printed out

In [25]:
# Question 5: check for missing values in df_ords
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

Question 5 Answer: there are 206,209 missing values in days_since_prior_order column; these could be first-time orders; therefore, there would be no days since prior order to record.

In [26]:
df_ords.shape

(3421083, 6)

In [27]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [28]:
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

Question 6 Answer: Exploring the data further, we can confirm that the missing values in days_since_prior_order are indeed first-time orders: there are 206,209 missing values in the column (NaN), as well as 206,209 unique user_id's. Therefore, the dataframe is made up of 206,209 customers - each with a first order (listed as NaN). I could change the missing values to '0'; but '0' already exists as a value in the column (67,755); this could be customers that ordered more than once in the same day. Therefore, I don't want to change the NaN values to zero, because I think it's important to distinguish between 1st-time orders and same-day orders. Also, I wouldn't delete these rows as they are valid orders and account for over 5% of the data (206,209 / 3,421,083 = 6%). I could impute a string '1st order' for these missing values. But then we would have mixed-type data in the column. Therefore, I have decided to leave these missing values as is. Elias, can you suggest a better option?

In [29]:
# Question 7: create new dataframe with duplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [30]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

Question 7 Answer: There are no duplicates in df_ords

In [31]:
# Export df_ords as is to 'Prepared Data' folder
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

In [32]:
# Replace NaN with -1, as per tutor suggestion

In [33]:
df_ords_clean = pd.DataFrame()

In [34]:
df_ords_clean = df_ords.fillna(-1)

In [35]:
df_ords_clean.isnull().sum()

order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

In [36]:
df_ords_clean['days_since_prior_order'].value_counts(dropna = False)

 30.0    369323
 7.0     320608
 6.0     240013
 4.0     221696
 3.0     217005
 5.0     214503
-1.0     206209
 2.0     193206
 8.0     181717
 1.0     145247
 9.0     118188
 14.0    100230
 10.0     95186
 13.0     83214
 11.0     80970
 12.0     76146
 0.0      67755
 15.0     66579
 16.0     46941
 21.0     45470
 17.0     39245
 20.0     38527
 18.0     35881
 19.0     34384
 22.0     32012
 28.0     26777
 23.0     23885
 27.0     22013
 24.0     20712
 25.0     19234
 29.0     19191
 26.0     19016
Name: days_since_prior_order, dtype: int64

In [37]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [38]:
df_ords_clean.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                    -1.0  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [39]:
# Export df_ords_clean to 'Prepared Data' folder
df_ords_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

In [46]:
df_ords_clean.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                    -1.0  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [41]:
df_ords_clean['days_since_prior_order'].value_counts(dropna = False)

 30.0    369323
 7.0     320608
 6.0     240013
 4.0     221696
 3.0     217005
 5.0     214503
-1.0     206209
 2.0     193206
 8.0     181717
 1.0     145247
 9.0     118188
 14.0    100230
 10.0     95186
 13.0     83214
 11.0     80970
 12.0     76146
 0.0      67755
 15.0     66579
 16.0     46941
 21.0     45470
 17.0     39245
 20.0     38527
 18.0     35881
 19.0     34384
 22.0     32012
 28.0     26777
 23.0     23885
 27.0     22013
 24.0     20712
 25.0     19234
 29.0     19191
 26.0     19016
Name: days_since_prior_order, dtype: int64

In [42]:
# Export df_ords_clean to 'Prepared Data' folder
df_ords_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))